pkg/cmd/httpd/server.go

```
@@ -0,0 +1,169 @@
+func ExecuteCmdHandler(cmd string, args []string, cmdExecutor *CmdExecutor) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var req struct {
+			Args []string               `json:"args"`
+			Env  map[string]interface{} `json:"env"`
+		}
+		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+			http.Error(w, "invalid body", http.StatusBadRequest)
+			return
+		}
+
+		cmd := exec.Command(cmd, append(args, req.Args...)...)
+		cmd.Env = os.Environ()
+		for k, v := range req.Env {
+			cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", k, v))
+		}
+		stdoutPipe, err := cmd.StdoutPipe()
+		if err != nil {
+			WriteHTTPError(w, r, err)
+			return
+		}
+		stderrPipe, err := cmd.StderrPipe()
+		if err != nil {
+			WriteHTTPError(w, r, err)
+			return
+		}
+
+		if err := cmdExecutor.Execute(cmd); err != nil {
+			WriteHTTPError(w, r, err)
+			return
+		}
+
+		flusher, ok := w.(http.Flusher)
+		if !ok {
+			WriteHTTPError(w, r, errors.New("expected http.ResponseWriter to be an http.Flusher"))
+			return
+		}
+
+		w.Header().Set("Connection", "Keep-Alive")
+		w.Header().Set("X-Content-Type-Options", "nosniff")
+		w.Header().Set("Transfer-Encoding", "chunked")
+		var mutex sync.Mutex
+
+		go WriteOutput(w, flusher, stdoutPipe, OutputTypeStdout, &mutex)
+		go WriteOutput(w, flusher, stderrPipe, OutputTypeStderr, &mutex)
+
+		exitStatus := cmdExecutor.Wait()
+		if err := json.NewEncoder(w).Encode(StatusOutput{
+			Output{
+				Msg:  exitStatus.Msg,
+				Type: OutputTypeStatus,
+			},
+			exitStatus.Code,
+		}); err != nil {
+			WriteHTTPError(w, r, err)
+		}
+	}
+}
+
+func WriteOutput(w http.ResponseWriter, flusher http.Flusher, pipe io.ReadCloser, outputType OutputType, writeMutex *sync.Mutex) {
+	scanner := bufio.NewScanner(pipe)
+	for scanner.Scan() {
+		output := Output{
+			Msg:  scanner.Text(),
+			Type: outputType,
+		}
+		writeMutex.Lock()
+		if err := json.NewEncoder(w).Encode(output); err != nil {
```

So here, `WriteOutput` is called once for stderr, and once for stdout. The behavior i want to achieve is 
1. If either `WriteOutput` fails, e.g. `Encode` or `scanner` have an error, i want the other `WriteOutput` to also stop. I tried doing this with a `context` but had trouble writing this function in such a way that a `<-ctx.Done()` can stop it.
2. if either `WriteOutput` fails, I want to surface that error back to the client. right now the exist status is coming from `exitStatus := cmdExecutor.Wait()`, so somehow i need to get this error there, or approach it a different way. one way i tried doing this was creating a new `chan error` that `WriteOutput` could write to and `Wait` would consume. is that the right way?
3. it would be cool if anytime the request context `r.Context` finishes or `WriteOutput` errors, `Wait` would call `Cancel()` to terminate the child process and then pass on the `r.Context`/`WriteOutput` error in the ExitStatus, not sure how to accomplish that though :/